doc/design.md

```diff
@@ -0,0 +1,922 @@
+- Store the full dimension extent in addition to the chunk extent in `Dimensions`.
+- If a dimension of a dataset represents just a subsection of the full dimension extent we use a special dimension tag.
+- Operations that operate on a particular dimension will then fail.
+  *Example: `Algorithms::DiffractionFocussing` requires `Dim::Spectrum`.*
+- Based on the stored full dimension extent we can assemble a dataset from chunks, or provide a multi-step operation that can deal with partial dimensions and does the assembling process.
+  *Example: A version of `Algorithms::DiffractionFocussing` that first focusses within each chunk and merges focussed data from all chunks in a final step.*
+
+
+### <a name="design-components-dimension-tags"></a>Dimension tags
+
+A dimension is identified by a tag.
+This is simply an `enum class Dim`.
+*Note:
+In the prototype this is currently named `enum class Dimension`, but we suggest renaming this to `enum class Dim` for brevity.
+Despite the abbreviation this should be clear and may actually improve code readability since it is more concise when used frequently?*
+
+An incomplete list of dimension tags:
+- `Dim::X`, `Dim::Y`, `Dim::Z`.
+- `Dim::Qx`, `Dim::Qy`, `Dim::Qz`.
+- `Dim::Detector`
+- `Dim::Spectrum`
+- `Dim::Tof`
+- `Dim::Wavelength`
+- `Dim::DetectorScan`
+- `Dim::Temperature`
+- `Dim::Polarization`
+- `Dim::Run`
+- `Dim::Row`
+- `Dim::Event`
+
+In many cases there is a corresponding coordinate tag for a dimension.
+
+
+### <a name="design-components-datasetview"></a>`class DatasetView`
+
+Due to the type-erasure we cannot simply treat `Dataset` as an array of structs, i.e., there is no equivalent to `API::MatrixWorkspace::getSpectrum(const size_t index)`.
+If only access to a single variable is required it can be access directly.
+For cases that require join access to multiple variables we provide `class DatasetView`:
+- Effectively provides a "zip-iterator", with support for automatic on-the-fly broadcast or transposition of dimensions.
+- Nested view allows for accessing a histogram-like view.
+- Allows for joint iteration with bin-edge variables using a `Bin` wrapper, overcoming the old nuisance of having `length+1` bin edges.
+
+The current implementation seems to support all required features.
+Performance was originally a problem due to the multi-dimensional iteration, but has been subsequently been optimized to the point where it should be sufficient for most applications.
+
+
+### <a name="design-components-datasetindex"></a>`class DatasetIndex`
+
+`class DatasetIndex` is a simple helper to index slices in a dataset via values in an axis (coordinate).
+`Dataset` itself makes no attempt at keeping values in a coordinate variable validated.
+For example, there is no check for duplicate spectrum numbers.
+Therefore, direct indexing is not possible.
+Instead `DatasetIndex` builds the index (and does the validation) on demand when constructed.
```

Comment:
If prototyped, can you point me to current prototype area that does this.

Reply:
https://github.com/mantidproject/dataset/blob/master/src/dataset_index.h